internal/pkg/siderolink/provision.go

```diff
@@ -0,0 +1,419 @@
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			var res resource.Resource
+
+			res, err = protobuf.CreateResource(rd.Type)
+			if err != nil {
+				return zero, err
+			}
+
+			*res.Metadata() = resource.NewMetadata(rd.DefaultNamespace, rd.Type, id, resource.VersionUndefined)
+
+			link, ok := res.(T)
+			if !ok {
+				return zero, fmt.Errorf("incorrect resource type")
+			}
+
+			if value, ok := provisionContext.token.ExtraData[omni.LabelInfraProviderID]; ok {
+				link.Metadata().Annotations().Set(omni.LabelInfraProviderID, value)
+			}
+
+			link.TypedSpec().Value, err = generateLinkSpec(provisionContext)
+			if err != nil {
+				return zero, err
+			}
+
+			if link.Metadata().Type() == siderolink.PendingMachineType {
+				link.Metadata().Annotations().Set("timestamp", time.Now().String())
+				link.Metadata().Labels().Set(omni.MachineUUID, provisionContext.request.NodeUuid)
+			}
+
+			return link, st.Create(ctx, res)
+		}
+
+		return zero, err
+	}
+
+	return r, nil
+}
+
+func (h *ProvisionHandler) updateLink(ctx context.Context, provisionContext *provisionContext) (*pb.ProvisionResponse, error) {
+	// if the token is not generated and the node supports secure join tokens
+	// put the machine into the limbo state by creating the pending machine resource
+	// the controller will then pick it up and create a wireguard peer for it
+	if pointer.SafeDeref(provisionContext.request.NodeUniqueToken) == "" && provisionContext.supportsSecureJoinTokens {
+		link, err := ensureResource[*siderolink.PendingMachine](ctx, h.state, provisionContext)
+		if err != nil {
+			return nil, err
+		}
+
+		if provisionContext.link == nil {
+			return h.genProvisionResponse(ctx, provisionContext, link, link.TypedSpec().Value)
+		}
+	}
+
+	link, err := ensureResource[*siderolink.Link](ctx, h.state, provisionContext)
+	if err != nil {
+		return nil, err
+	}
+
+	if link, err = safe.StateUpdateWithConflicts(ctx, h.state, link.Metadata(), func(r *siderolink.Link) error {
+		s := r.TypedSpec().Value
+
+		s.NodePublicKey = provisionContext.request.NodePublicKey
+		s.VirtualAddrport, err = generateVirtualAddrPort(pointer.SafeDeref(provisionContext.request.WireguardOverGrpc))
+		if err != nil {
+			return err
+		}
+
+		if s.NodeUniqueToken == "" && provisionContext.request.NodeUniqueToken != nil {
+			s.NodeUniqueToken = *provisionContext.request.NodeUniqueToken
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return h.genProvisionResponse(ctx, provisionContext, link, link.TypedSpec().Value)
+}
+
+func (h *ProvisionHandler) removePendingMachine(ctx context.Context, pendingMachine *siderolink.PendingMachine) error {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*5)
+	defer cancel()
+
+	_, err := h.state.Teardown(ctx, pendingMachine.Metadata())
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			return nil
+		}
+
+		return nil
+	}
+
+	_, err = h.state.WatchFor(ctx, pendingMachine.Metadata(), state.WithFinalizerEmpty())
```

Comment:
Drop the watch for, run teardown and destroy for the ready machines, then on the next loop run teardown again and destroy ready ones.